From 6369249f2ae1fff4a9862abc97aafc1b9d97473b Mon Sep 17 00:00:00 2001
From: Marc Zyngier <maz@kernel.org>
Date: Tue, 9 Apr 2024 09:22:10 +0800
Subject: [PATCH 585/960] downstream: arm64: Decouple KVM from CONFIG_ARM64_NMI

Even if the host isn't using NMIs, that isn't a reason for
preventing the feature being available to guests,

Decouple the two and let the capability be available irrespective
of the host NMI level of support.

Signed-off-by: Marc Zyngier <maz@kernel.org>
Signed-off-by: Xiang Chen <chenxiang66@hisilicon.com>
Signed-off-by: caijian <caijian11@h-partners.com>
Signed-off-by: chenyi <chenyi211@h-partners.com>
Conflicts:
	arch/arm64/kernel/cpufeature.c
[ Context conflicts. ]
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 arch/arm64/include/asm/cpufeature.h |  3 ++-
 arch/arm64/kernel/cpufeature.c      | 13 +++++++++----
 arch/arm64/kvm/arm.c                |  2 +-
 arch/arm64/kvm/sys_regs.c           |  3 ++-
 drivers/irqchip/irq-gic-v3.c        | 13 ++++++++++++-
 5 files changed, 26 insertions(+), 8 deletions(-)

diff --git a/arch/arm64/include/asm/cpufeature.h b/arch/arm64/include/asm/cpufeature.h
index 5dc0a91c2616..7c8401b4b78c 100644
--- a/arch/arm64/include/asm/cpufeature.h
+++ b/arch/arm64/include/asm/cpufeature.h
@@ -808,7 +808,8 @@ static __always_inline bool system_uses_irq_prio_masking(void)
 static __always_inline bool system_uses_nmi(void)
 {
 	return IS_ENABLED(CONFIG_ARM64_NMI) &&
-		cpus_have_final_cap(ARM64_USES_NMI);
+		cpus_have_final_cap(ARM64_USES_NMI) &&
+		!system_uses_irq_prio_masking();
 }
 
 static inline bool system_supports_mte(void)
diff --git a/arch/arm64/kernel/cpufeature.c b/arch/arm64/kernel/cpufeature.c
index b06d65f1f120..0cc324cf3584 100644
--- a/arch/arm64/kernel/cpufeature.c
+++ b/arch/arm64/kernel/cpufeature.c
@@ -2242,20 +2242,24 @@ static bool has_gic_prio_relaxed_sync(const struct arm64_cpu_capabilities *entry
 }
 #endif
 
-#ifdef CONFIG_ARM64_NMI
 static bool use_nmi(const struct arm64_cpu_capabilities *entry, int scope)
 {
 	if (!has_cpuid_feature(entry, scope))
 		return false;
 
 	/*
+	 * NMI support was not enabled in the kernel, but can still be
+	 * used by guests. Let the world know.
+	 *
 	 * Having both real and pseudo NMIs enabled simultaneously is
 	 * likely to cause confusion.  Since pseudo NMIs must be
 	 * enabled with an explicit command line option, if the user
 	 * has set that option on a system with real NMIs for some
 	 * reason assume they know what they're doing.
 	 */
-	if (IS_ENABLED(CONFIG_ARM64_PSEUDO_NMI) && enable_pseudo_nmi) {
+	if (!IS_ENABLED(CONFIG_ARM64_NMI))
+		pr_info("CONFIG_ARM64_NMI disabled, using NMIs for guests only\n");
+	else if (IS_ENABLED(CONFIG_ARM64_PSEUDO_NMI) && enable_pseudo_nmi) {
 		pr_info("Pseudo NMI enabled, not using architected NMI\n");
 		return false;
 	}
@@ -2263,6 +2267,7 @@ static bool use_nmi(const struct arm64_cpu_capabilities *entry, int scope)
 	return true;
 }
 
+#ifdef CONFIG_ARM64_NMI
 static void nmi_enable(const struct arm64_cpu_capabilities *__unused)
 {
 	/*
@@ -2895,7 +2900,6 @@ static const struct arm64_cpu_capabilities arm64_features[] = {
 		.cpu_enable = cpu_enable_ls64_v,
 		ARM64_CPUID_FIELDS(ID_AA64ISAR1_EL1, LS64, LS64_V)
 	},
-#ifdef CONFIG_ARM64_NMI
 	{
 		.desc = "Non-maskable Interrupts present",
 		.capability = ARM64_HAS_NMI,
@@ -2917,9 +2921,10 @@ static const struct arm64_cpu_capabilities arm64_features[] = {
 		.field_width = 4,
 		.min_field_value = ID_AA64PFR1_EL1_NMI_IMP,
 		.matches = use_nmi,
+#ifdef CONFIG_ARM64_NMI
 		.cpu_enable = nmi_enable,
-	},
 #endif
+	};
 #ifdef CONFIG_ARM64_TWED
 	{
 		.desc = "Delayed Trapping of WFE",
diff --git a/arch/arm64/kvm/arm.c b/arch/arm64/kvm/arm.c
index 415e3914ace7..bb994ed76c04 100644
--- a/arch/arm64/kvm/arm.c
+++ b/arch/arm64/kvm/arm.c
@@ -306,7 +306,7 @@ int kvm_arch_init_vm(struct kvm *kvm, unsigned long type)
 	/* The maximum number of VCPUs is limited by the host's GIC model */
 	kvm->max_vcpus = kvm_arm_default_max_vcpus();
 
-	if (system_uses_nmi() && !static_branch_unlikely(&vgic_v3_cpuif_trap))
+	if (cpus_have_final_cap(ARM64_HAS_NMI) && !static_branch_unlikely(&vgic_v3_cpuif_trap))
 		kvm->arch.pfr1_nmi = ID_AA64PFR1_EL1_NMI_IMP;
 
 	kvm_arm_init_hypercalls(kvm);
diff --git a/arch/arm64/kvm/sys_regs.c b/arch/arm64/kvm/sys_regs.c
index 07d0f27ffdad..f1cfe4004881 100644
--- a/arch/arm64/kvm/sys_regs.c
+++ b/arch/arm64/kvm/sys_regs.c
@@ -1624,7 +1624,8 @@ static int set_id_aa64pfr1_el1(struct kvm_vcpu *vcpu,
 
 	nmi = cpuid_feature_extract_unsigned_field(val, ID_AA64PFR1_EL1_NMI_SHIFT);
 	if (nmi > ID_AA64PFR1_EL1_NMI_IMP ||
-	    (nmi && (!system_uses_nmi() || static_branch_unlikely(&vgic_v3_cpuif_trap))))
+	    (nmi && (!cpus_have_final_cap(ARM64_HAS_NMI) ||
+		     static_branch_unlikely(&vgic_v3_cpuif_trap))))
 		return -EINVAL;
 
 	/* We can only differ with NMI, and anything else is an error */
diff --git a/drivers/irqchip/irq-gic-v3.c b/drivers/irqchip/irq-gic-v3.c
index cc1b9f1d525a..735024213ac7 100644
--- a/drivers/irqchip/irq-gic-v3.c
+++ b/drivers/irqchip/irq-gic-v3.c
@@ -151,11 +151,21 @@ static inline bool has_v3_3_nmi(void)
 {
 	return gic_data.has_nmi && system_uses_nmi();
 }
+
+static bool system_is_nmi_capable(void)
+{
+	return gic_data.has_nmi && cpus_have_final_cap(ARM64_HAS_NMI);
+}
 #else
 static inline bool has_v3_3_nmi(void)
 {
 	return false;
 }
+
+static bool system_is_nmi_capable(void)
+{
+	return false;
+}
 #endif
 
 #ifdef CONFIG_VIRT_VTIMER_IRQ_BYPASS
@@ -2368,6 +2378,7 @@ static void __init gic_of_setup_kvm_info(struct device_node *node)
 
 	gic_v3_kvm_info.has_v4 = gic_data.rdists.has_vlpis;
 	gic_v3_kvm_info.has_v4_1 = gic_data.rdists.has_rvpeid;
+	gic_v3_kvm_info.has_nmi = system_is_nmi_capable();
 #ifdef CONFIG_VIRT_VTIMER_IRQ_BYPASS
 	gic_v3_kvm_info.has_vtimer = gic_data.rdists.has_vtimer;
 #endif
@@ -2711,7 +2722,7 @@ static void __init gic_acpi_setup_kvm_info(void)
 
 	gic_v3_kvm_info.has_v4 = gic_data.rdists.has_vlpis;
 	gic_v3_kvm_info.has_v4_1 = gic_data.rdists.has_rvpeid;
-	gic_v3_kvm_info.has_nmi = has_v3_3_nmi();
+	gic_v3_kvm_info.has_nmi = system_is_nmi_capable();
 #ifdef CONFIG_VIRT_VTIMER_IRQ_BYPASS
 	gic_v3_kvm_info.has_vtimer = gic_data.rdists.has_vtimer;
 #endif
-- 
2.43.0

